pkg/trustedresources/verify.go

```diff
@@ -169,32 +175,68 @@ func getMatchedPolicies(resourceName string, source string, policies []*v1alpha1
-func verifyResource(ctx context.Context, resource metav1.Object, k8s kubernetes.Interface, signature []byte, matchedPolicies []*v1alpha1.VerificationPolicy) error {
+//
+// TODO(#6683): return all failed policies in error.
+func verifyResource(ctx context.Context, resource metav1.Object, k8s kubernetes.Interface, signature []byte, matchedPolicies []*v1alpha1.VerificationPolicy) VerificationResult {
+	logger := logging.FromContext(ctx)
+	var warnPolicies []*v1alpha1.VerificationPolicy
+	var enforcePolicies []*v1alpha1.VerificationPolicy
 	for _, p := range matchedPolicies {
+		if p.Spec.Mode == v1alpha1.ModeWarn {
+			warnPolicies = append(warnPolicies, p)
+		} else {
+			enforcePolicies = append(enforcePolicies, p)
+		}
+	}
+
+	// first evaluate all enforce policies. Return VerificationError type of VerificationResult if any policy fails.
+	for _, p := range enforcePolicies {
 		passVerification := false
 		verifiers, err := verifier.FromPolicy(ctx, k8s, p)
 		if err != nil {
-			return fmt.Errorf("failed to get verifiers from policy: %w", err)
+			return VerificationResult{VerificationResultType: VerificationError, Err: fmt.Errorf("failed to get verifiers from policy: %w", err)}
 		}
-		for _, verifier := range verifiers {
-			// if one of the verifier passes verification, then this policy passes verification
-			if err := verifyInterface(resource, verifier, signature); err == nil {
-				passVerification = true
-				break
-			}
+		if pass := doesAnyVerifierPass(resource, signature, verifiers); pass {
+			passVerification = pass
 		}
-		// if this policy fails the verification and the mode is not "warn", should return error directly. No need to check other policies
 		if !passVerification {
-			if p.Spec.Mode == v1alpha1.ModeWarn {
-				logger := logging.FromContext(ctx)
-				logger.Warnf("%w: resource %s in namespace %s fails verification", ErrResourceVerificationFailed, resource.GetName(), resource.GetNamespace())
-			} else {
-				// if the mode is "enforce" or not set, return error.
-				return fmt.Errorf("%w: resource %s in namespace %s fails verification", ErrResourceVerificationFailed, resource.GetName(), resource.GetNamespace())
-			}
+			return VerificationResult{VerificationResultType: VerificationError, Err: fmt.Errorf("%w: resource %s in namespace %s fails verification", ErrResourceVerificationFailed, resource.GetName(), resource.GetNamespace())}
 		}
 	}
-	return nil
+
+	// return VerificationWarn type of VerificationResult if any warn policies fails.
+	for _, p := range warnPolicies {
+		passVerification := false
+		verifiers, err := verifier.FromPolicy(ctx, k8s, p)
+		if err != nil {
+			warn := fmt.Errorf("fails to get verifiers for resource %s from namespace %s: %w", resource.GetName(), resource.GetNamespace(), err)
+			logger.Warnf(warn.Error())
+			return VerificationResult{VerificationResultType: VerificationWarn, Err: warn}
+		}
+		if pass := doesAnyVerifierPass(resource, signature, verifiers); pass {
```

Comment:
why create the variables `pass` and `passverification`? do they have different purposes?

Reply:
oh yes, thanks! That was a mistake 😞